pallets/usage/src/lib.rs

```diff
@@ -343,38 +344,42 @@ pub mod pallet {
                                 let trap_message_id = core_processor::next_message_id(
                                     program_id.as_ref().into(),
                                     program.nonce,
-                                );
+                                ).into_origin();
                                 let trap_message = Message {
-                                    id: H256::from_slice(trap_message_id.as_slice()),
+                                    id: trap_message_id,
                                     source: program_id,
-                                    dest: msg.source,
+                                    dest: message.source,
                                     payload: vec![],
                                     gas_limit: trap_gas,
                                     value: 0,
-                                    reply: Some((msg.id, core_processor::ERR_EXIT_CODE)),
+                                    reply: Some((message.id, core_processor::ERR_EXIT_CODE)),
                                 };
 
+                                let dispatch = Dispatch::handle_reply(trap_message);
+
                                 // Enqueue the trap reply message
-                                let _ = gas_tree.split_off(trap_message.id, trap_gas);
-                                common::queue_message(trap_message);
+                                let _ = gas_tree.split_off(trap_message_id, trap_gas);
+                                common::queue_dispatch(dispatch);
 
                                 // Save back the program with incremented nonce
                                 common::set_program(program_id, program, Default::default());
                             }
                             _ => {
-                                // Must be unreachable: there shouldn't be dangling messages in the WL
-                                // if the associated program has been killed.
-                                // TODO: ensure the above statement always holds:
-                                // https://github.com/gear-tech/gear/issues/507
-                                log::warn!("Failed to find a program with ID: {:?}", program_id);
+                                unreachable!(
+                                    "program with {:?} id was terminated and messages to it were remove from WL",
```

Comment:
Do we remove them atm?

Reply:
yes, when program was terminated, messaged to it are removed from WL in manager